src/ConnectedMode/BranchMatcher.cs

```diff
@@ -0,0 +1,119 @@
+        /// Calculates the nearest branch to the head branch.
+        /// Returns null if there are no head branches
+        /// </summary>
+        /// <remarks>
+        /// Compares branches by name and calculates distances by the number of different commits.
+        /// Defaults to branch marked as "main" when no match found.
+        /// This is the same algorithm as the other SonarLint flavours (?except in how we treated branch histories.
+        /// We treat branch histories as series of linear commits ordered by time, even if there are merges).
+        /// </remarks>
+        Task<string> GetMatchedBranch(string projectKey);
+    }
+
+    internal class BranchMatcher : IBranchMatcher
+    {
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly IRepository gitRepo;
+
+        public BranchMatcher(ISonarQubeService sonarQubeService, IRepository gitRepo)
+        {
+            this.sonarQubeService = sonarQubeService;
+            this.gitRepo = gitRepo;
+        }
+
+        public async Task<string> GetMatchedBranch(string projectKey)
+        {
+            Debug.Assert(sonarQubeService.IsConnected,
+                "Not expecting GetMatchedBranch to be called unless we are in Connected Mode");
+
+            var head = gitRepo.Head;
+
+            if (head == null)
+            {
+                return null;
+            }
+
+            var remoteBranches = await sonarQubeService.GetProjectBranchesAsync(projectKey, CancellationToken.None);
+
+            if (remoteBranches.Any(rb => string.Equals(rb.Name, head.FriendlyName, StringComparison.InvariantCultureIgnoreCase)))
+            {
+                return head.FriendlyName;
+            }
+
+            string closestBranch = null;
+            int closestDistance = int.MaxValue;
+
+            foreach (var remoteBranch in remoteBranches)
+            {
+                var localBranch = gitRepo.Branches.FirstOrDefault(r => string.Equals(r.FriendlyName, remoteBranch.Name, StringComparison.InvariantCultureIgnoreCase));
+
+                if (localBranch == null) { continue; }
+
+                var distance = GetDistance(head, localBranch);
+
+                if (distance < closestDistance)
+                {
+                    closestBranch = localBranch.FriendlyName;
+                    closestDistance = distance;
+                }
+            }
+
+            if (closestBranch == null)
+            {
+                return remoteBranches.First(rb => rb.IsMain).Name;
+            }
+            return closestBranch;
+        }
+
+        //Commits are in descending order by time
+        private int GetDistance(Branch head, Branch branch)
+        {
+            var headCommits = head.Commits.ToList();
+            for (int i = 0; i < headCommits.Count; i++)
+            {
+                var commitID = headCommits[i].Id;
+
+                var branchCommitIndex = branch.Commits.ToList().FindIndex(bc => bc.Id == commitID);
```

Comment:
`ToList` means we're going to enumerate the commit history for the branch, even if the commit we want is first in the list. We're going to repeat that for every branch.
I'll add a TODO to look at optimising this.